### PR TITLE
[C3] Use wrangler.toml for deployment in full-stack templates

### PR DESCRIPTION
**Pull request by jculvey**
## What this PR solves / how to test

Updates the full-stack templates to use `pages_build_output_directory` so that `dev` and `deploy` commands are driven off of `wrangler.toml`.

## Author has addressed the following

- Tests
  - [ ] TODO (before merge)
  - [ ] Included
  - [x] Not necessary because: Existing coverage sufficient
- Changeset ([Changeset guidelines](https://github.com/cloudflare/workers-sdk/blob/main/CONTRIBUTING.md#changesets))
  - [ ] TODO (before merge)
  - [ ] Included
  - [x] Not necessary because:
- Public documentation
  - [ ] TODO (before merge)
  - [ ] Cloudflare docs PR(s): <https://github.com/cloudflare/cloudflare-docs/pull/>...
  - [x] Not necessary because: Features already documented

### CODE REVIEW

File: packages/create-cloudflare/templates/astro/templates/wrangler.toml [hunk 1]
```diff
@@ -1,5 +1,11 @@
 name = "<TBD>"
 compatibility_date = "<TBD>"
+pages_build_output_dir = "./dist"
+
+# Automatically places your application in an optimal location that minimizes latency
+# Docs: https://developers.cloudflare.com/pages/functions/smart-placement/
+[placement]
+mode = "smart"
```
Comment (IgorMinar) [inline packages/create-cloudflare/templates/astro/templates/wrangler.toml:8]: 🎉

**Review by Cherry — COMMENTED** [review]
It's awesome to see additional bindings, and the pages projects using `pages_build_output_dir` by default, but I do not think Smart Placement should be enabled by default for a few reasons:

- Smart Placement is still beta. A template like this should in my opinion really only enable features that are battle-tested, and ready for production use.

- It kinda goes against a lot of the marketing people have been given for years about Workers/Pages, with their code running "everywhere", "300+ cities", etc. With Smart Placement enabled, this might no longer be true, depending on what your Worker is doing.

- Smart Placement still requires a pretty specific build out to be beneficial, with splitting up services into their own workers. https://developers.cloudflare.com/workers/configuration/smart-placement/#best-practices covers this a little.

- It's not usable at all yet with other powerful features, like Gradual Deployments.

- And finally, Smart Placement still has a number of issues that make it not ready for production traffic and enabled by default
    - Every week in Discord, we see many users hit random 500s with Smart Placement enabled that get resolved as soon as they turn Smart Placement off and redeploy
    - Weird obscure issues like with Durable Object placement that get resolved when Smart Placement is disabled
    - And Smart Placement still completely implodes Pages projects if you try to `tail` them, while using `env.ASSETS`. This is unrecoverable, and the user must go re-deploy their project entirely.


Please reconsider enabling Smart Placement by default. It should be an opt-in feature once people understand exactly what it does and how their worker will operate once it is turned on.

**Comment by jculvey** [comment]
> It's awesome to see additional bindings, and the pages projects using `pages_build_output_dir` by default, but I do not think Smart Placement should be enabled by default for a few reasons:
> 

I've removed the changes to turn on Smart Placement by default.

File: packages/create-cloudflare/templates/astro/templates/wrangler.toml [hunk 2]
```diff
@@ -49,4 +50,22 @@ compatibility_date = "<TBD>"
...
 # binding = "MY_SERVICE"
-# service = "my-service"
\ No newline at end of file
+# service = "my-service"
+
+# To use different bindings for preview and production environments, follow the examples below.
+# When using environment-specific overrides for bindings, ALL bindings must be specified on a per-environment basis.
+# Docs: https://developers.cloudflare.com/pages/functions/wrangler-configuration#environment-specific-overrides
+
+# [env.preview.vars]
```
Comment (IgorMinar) [inline packages/create-cloudflare/templates/astro/templates/wrangler.toml:59]: Consider adding divider like the following to make it easier to scan the file

```suggestion
######## PREVIEW environment config ########

# [env.preview.vars]
```

File: packages/create-cloudflare/templates/astro/templates/wrangler.toml [hunk 3]
```diff
@@ -49,4 +50,22 @@ compatibility_date = "<TBD>"
...
+# API_KEY = "xyz789"
+
+# [[env.preview.kv_namespaces]]
+# binding = "MY_KV_NAMESPACE"
+# id = "<PREVIEW_NAMESPACE_ID>"
+
+# [env.production.vars]
```
Comment (IgorMinar) [inline packages/create-cloudflare/templates/astro/templates/wrangler.toml:66]:
```suggestion
######## PRODUCTION environment config ########

# [env.production.vars]
```

**Review by IgorMinar — COMMENTED** [review]
can you please add solid as well?

File: .changeset/two-phones-drop.md [hunk 4]
```diff
@@ -0,0 +1,8 @@
+---
+"create-cloudflare": minor
+---
+
+feat: Update full-stack templates to leverage newly released features from dev week.
+
+- `wrangler.toml` will now be used for for deployment and previews.
+- Smart placement will be enabled by default.
```
Comment (Cherry) [inline .changeset/two-phones-drop.md:8]: nitpick: the smart placement line should probably be removed for now so this doesn't land in the changelog

Comment (jculvey) [reply .changeset/two-phones-drop.md:8]: Good catch, thanks 😄

File: packages/create-cloudflare/templates/astro/templates/wrangler.toml [hunk 5]
```diff
@@ -50,4 +51,26 @@ compatibility_date = "<TBD>"
...
+# id = "<PREVIEW_NAMESPACE_ID>"
+
+######## PRODUCTION environment config ########
+
+# [env.production.vars]
+# API_KEY = "abc123"
+
+# [[env.production.kv_namespaces]]
+# binding = "MY_KV_NAMESPACE"
+# id = "<PRODUCTION_NAMESPACE_ID>"
```
Comment (dario-piotrowicz) [inline packages/create-cloudflare/templates/astro/templates/wrangler.toml:76]:
```suggestion
# id = "<PRODUCTION_NAMESPACE_ID>"

```

File: packages/create-cloudflare/templates/hello-world-durable-object/ts/wrangler.toml [hunk 6]
```diff
@@ -95,5 +95,4 @@ new_classes = ["MyDurableObject"]
 # Docs: https://developers.cloudflare.com/workers/wrangler/configuration/#vectorize-indexes
 # [[vectorize]]
 # binding = "MY_INDEX"
-# index_name = "my-index"
-
+# index_name = "my-index"
```
Comment (dario-piotrowicz) [inline packages/create-cloudflare/templates/hello-world-durable-object/ts/wrangler.toml:98]:
```suggestion
# index_name = "my-index"

```

File: packages/create-cloudflare/templates/next/wrangler.toml [hunk 7]
```diff
@@ -53,3 +54,24 @@ compatibility_flags = ["nodejs_compat"]
...
+# id = "<PREVIEW_NAMESPACE_ID>"
+
+######## PRODUCTION environment config ########
+
+# [env.production.vars]
+# API_KEY = "abc123"
+
+# [[env.production.kv_namespaces]]
+# binding = "MY_KV_NAMESPACE"
+# id = "<PRODUCTION_NAMESPACE_ID>"
```
Comment (dario-piotrowicz) [inline packages/create-cloudflare/templates/next/wrangler.toml:77]:
```suggestion
# id = "<PRODUCTION_NAMESPACE_ID>"

```

File: packages/create-cloudflare/templates/nuxt/templates/wrangler.toml [hunk 8]
```diff
@@ -50,4 +51,26 @@ compatibility_date = "<TBD>"
...
+# id = "<PREVIEW_NAMESPACE_ID>"
+
+######## PRODUCTION environment config ########
+
+# [env.production.vars]
+# API_KEY = "abc123"
+
+# [[env.production.kv_namespaces]]
+# binding = "MY_KV_NAMESPACE"
+# id = "<PRODUCTION_NAMESPACE_ID>"
```
Comment (dario-piotrowicz) [inline packages/create-cloudflare/templates/nuxt/templates/wrangler.toml:76]:
```suggestion
# id = "<PRODUCTION_NAMESPACE_ID>"

```

File: packages/create-cloudflare/templates/qwik/templates/wrangler.toml [hunk 9]
```diff
@@ -50,4 +51,26 @@ compatibility_date = "<TBD>"
...
+# id = "<PREVIEW_NAMESPACE_ID>"
+
+######## PRODUCTION environment config ########
+
+# [env.production.vars]
+# API_KEY = "abc123"
+
+# [[env.production.kv_namespaces]]
+# binding = "MY_KV_NAMESPACE"
+# id = "<PRODUCTION_NAMESPACE_ID>"
```
Comment (dario-piotrowicz) [inline packages/create-cloudflare/templates/qwik/templates/wrangler.toml:76]:
```suggestion
# id = "<PRODUCTION_NAMESPACE_ID>"

```

File: packages/create-cloudflare/templates/remix/templates/wrangler.toml [hunk 10]
```diff
@@ -50,4 +51,26 @@ compatibility_date = "<TBD>"
...
+# id = "<PREVIEW_NAMESPACE_ID>"
+
+######## PRODUCTION environment config ########
+
+# [env.production.vars]
+# API_KEY = "abc123"
+
+# [[env.production.kv_namespaces]]
+# binding = "MY_KV_NAMESPACE"
+# id = "<PRODUCTION_NAMESPACE_ID>"
```
Comment (dario-piotrowicz) [inline packages/create-cloudflare/templates/remix/templates/wrangler.toml:76]:
```suggestion
# id = "<PRODUCTION_NAMESPACE_ID>"

```

File: packages/create-cloudflare/templates/solid/templates/wrangler.toml [hunk 11]
```diff
@@ -0,0 +1,77 @@
...
+# id = "<PREVIEW_NAMESPACE_ID>"
+
+######## PRODUCTION environment config ########
+
+# [env.production.vars]
+# API_KEY = "abc123"
+
+# [[env.production.kv_namespaces]]
+# binding = "MY_KV_NAMESPACE"
+# id = "<PRODUCTION_NAMESPACE_ID>"
```
Comment (dario-piotrowicz) [inline packages/create-cloudflare/templates/solid/templates/wrangler.toml:77]:
```suggestion
# id = "<PRODUCTION_NAMESPACE_ID>"

```

File: packages/create-cloudflare/templates/svelte/js/wrangler.toml [hunk 12]
```diff
@@ -50,4 +51,26 @@ compatibility_date = "<TBD>"
...
+# id = "<PREVIEW_NAMESPACE_ID>"
+
+######## PRODUCTION environment config ########
+
+# [env.production.vars]
+# API_KEY = "abc123"
+
+# [[env.production.kv_namespaces]]
+# binding = "MY_KV_NAMESPACE"
+# id = "<PRODUCTION_NAMESPACE_ID>"
```
Comment (dario-piotrowicz) [inline packages/create-cloudflare/templates/svelte/js/wrangler.toml:76]:
```suggestion
# id = "<PRODUCTION_NAMESPACE_ID>"

```

File: packages/create-cloudflare/templates/svelte/ts/wrangler.toml [hunk 13]
```diff
@@ -50,4 +51,26 @@ compatibility_date = "<TBD>"
...
+# id = "<PREVIEW_NAMESPACE_ID>"
+
+######## PRODUCTION environment config ########
+
+# [env.production.vars]
+# API_KEY = "abc123"
+
+# [[env.production.kv_namespaces]]
+# binding = "MY_KV_NAMESPACE"
+# id = "<PRODUCTION_NAMESPACE_ID>"
```
Comment (dario-piotrowicz) [inline packages/create-cloudflare/templates/svelte/ts/wrangler.toml:76]:
```suggestion
# id = "<PRODUCTION_NAMESPACE_ID>"

```

**Review by petebacondarwin — APPROVED** [review]
Nice!

File: packages/create-cloudflare/templates/next/wrangler.toml [hunk 14]
```diff
@@ -53,3 +54,24 @@ compatibility_flags = ["nodejs_compat"]
...
+# Docs: https://developers.cloudflare.com/pages/functions/wrangler-configuration#environment-specific-overrides
+
+######## PREVIEW environment config ########
+
+# [env.preview.vars]
+# API_KEY = "xyz789"
+
+# [[env.preview.kv_namespaces]]
+# binding = "MY_KV_NAMESPACE"
+# id = "<PREVIEW_NAMESPACE_ID>"
```
Comment (dario-piotrowicz) [inline packages/create-cloudflare/templates/next/wrangler.toml:68]: Everywhere else we present examples of KVs, as their id we use: `# id = "xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx"`
I don't like that this is inconsistent with that 😅 

I can see why you're doing it (to distinguish between preview and prod ids)....

I don't think this needs to block this PR, but I would suggest to maybe, as a followup replace all the `xxxx` values we've been using with this sort of thing? (e.g.`<NAMESPACE_ID>`, `<DATABASE_ID>`, etc...)

**Review by dario-piotrowicz — APPROVED** [review]

File: packages/create-cloudflare/templates/next/wrangler.toml (hunk 14 above)
Comment (jculvey) [reply packages/create-cloudflare/templates/next/wrangler.toml:68]: This doesn't really bug me, and I think there's value in it. For example, the placeholder we use for D1 ids is: `"xxxxxxxx-xxxx-xxxx-xxxx-xxxxxxxxxxxx"`, which gives the user a little extra affordance that what they're copying is the correct thing.

It's ok to be inconsistent at times when there's a good reason for it 😄.

**Pull request merged**